Add Twilio Video Integration to Use Cases

digitalfuel opened the pull request.

digitalfuel committed:
Add Twilio Video Integration to Use Cases

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+
+## Problem/situation
+
+First, we needed to get and store all of the API Keys necessary to authorize the connection. We could get some of them from the platformOS Partner Portal after adding the Twilio integration as detailed in the topic [Managing Integrations](https://documentation.platformos.com/tutorials/partner-portal/instances/managing-integrations). To use other API resources, we created a form and listened to pO ...
```
```suggestion
First, we needed to get and store all of the API Keys necessary to authorize the connection. We could get some of them from the platformOS Partner Portal after adding the Twilio integration as detailed in the topic [Managing Integrations](/tutorials/partner-portal/instances/managing-integrations). To use other API resources, we created a form and listened to pOS logs in our CLI. Then we created a JWT (JSON Web Token) that would allow the connection to take place. 
```

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+
+#### Create API Key Secret with Curl
+Simply replace **ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX** with your Twilio `sid` and your `auth_token` from the [platformOS Partner Portal](https://portal.apps.near-me.com/): select the Instance your are working on, click Update configuration and then Edit to view the Twilio API Keys.
+<pre class="command-line" data-output="2-50" data-user="user" data-host="host"><code class="language-bash">curl -X POST https://api.twilio.com/2010-04-01/Accounts/ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX/Keys.json \ -u ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:your_auth_token</code></pre>
```
```suggestion
<pre class="command-line" data-output="2-50" data-user="user" data-host="host"><code class="language-bash">
curl -X POST https://api.twilio.com/2010-04-01/Accounts/ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX/Keys.json \ -u ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:your_auth_token
</code></pre>
```

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+To do this, we needed to connect directly to Twilio from platformOS.
+
+#### Create API Key Secret with platformOS - GraphQL GUI 
+Here, we created an `api_call_notification` and then used the `pos_cli gui sevre <environment>` CLI command to send and receive the response.
```
```suggestion
Here, we created an `api_call_notification` and then used the `pos-cli gui serve <environment>` CLI command to send and receive the response.
```

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+    headers: |-
+    {%- assign authBasic = context.constants.TWILIO_ACCOUNT_SID | append: ":" | append: context.constants.TWILIO_ACCOUNT_AUTH_TOKEN | base64_encode -%}
+        {
+            "Content-Type": "application/json; charset=utf-8",
```
I think charset part is not necessary :)

digitalfuel replied:
Yes you right on this. I can't remember exactly why I added it, it was a while back. I think I was having issues and getting an error code from Twilio. Adding it seemed to fix things. But I just tested it and it still works so not sure what happened before. I will remove.

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+
+2. **Send and receive the response**
+
+    To send the **API call notification** and receive the response we used the `pos_cli gui serve <environment>` command. We saw the response returned with our API Key **Secret**.
```
```suggestion
    To send the **API call notification** and receive the response we used the `pos-cli gui serve <environment>` command. We saw the response returned with our API Key **Secret**.
```

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
````diff
@@ -0,0 +1,302 @@
+
+    ```graphql
+    mutation createSecret {
+        api_call_send(template: {name: "twilio_api_secret_create"}) {
````
```suggestion
        api_call_send(template: { name: "twilio_api_secret_create" }) {
```

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+1. **Create a page**
+
+    First, we needed a page to hold a form.
+
```
Would be nice to add path to the file just before code starts.

digitalfuel replied:
Good suggestion. I have added, `#### views/pages/twilio-api.html.liquid`. If this is not what you meant let me know.

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
````diff
@@ -0,0 +1,302 @@
+    
+    ```liquid
+    {% raw %}
+    h3>Get Keys</h3>
````
```suggestion
    <h3>Get Keys</h3>
```

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+
+3. **Create a form**
+
+    The simpliest of forms, a button with a resource of our virtual_model schema and an `api_call_notification` that will be triggered on submit.
```
```suggestion
    The simplest of forms, a button with a resource of our `virtual_model` schema and an `api_call_notification` that will be triggered on submit.
```

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+    to: |-
+        https://api.twilio.com/2010-04-01/Accounts/{{ context.constants.TWILIO_ACCOUNT_SID }}/Keys/{{ context.constants.TWILIO_API_KEY }}.json
+    callback: |-
+        {%- log response.body, type: 'info' -%}
```
I would add a note that its good to remove logs from production code, to not create noise in logs.

Or if you need logs, i would prefix it with a module name instead of "info" - there might be a lot of info from a lot of different modules at some point in the future.

digitalfuel replied:
Good point. I have added a alert/note for this.

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+
+5. **View the response**
+
+    We clicked the button that submits the form that makes an **API call notification** but what happened to our **API Call Response**? No worries, we just needed to go to the CLI and use the `pos_cli logs <environment>` command to see the response returned as type `info` that we declared in the callback above.
```
```suggestion
    We clicked the button that submits the form that makes an **API call notification** but what happened to our **API Call Response**? No worries, we just needed to go to the CLI and use the `pos-cli logs <environment>` command to see the response returned as type `info` that we declared in the callback above.
```


If i miss pos_cli somewhere in the article, just know that its pos-cli :)

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+
+### Step 2: Store the API Keys with custom `context.constants`
+
+We created custom `context.constants` for each of the keys we need to store and access securely. To do this we used the [GraphQL GUI](https://documentation.platformos.com/best-practices/qa/developing-graphql-queries-using-pos-cli-gui) with the `pos_cli gui serve` command.
```
```suggestion
We created custom `context.constants` for each of the keys we need to store and access securely. To do this we used the [GraphQL GUI](/best-practices/qa/developing-graphql-queries-using-pos-cli-gui) with the `pos-cli gui serve` command.
```

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
````diff
@@ -0,0 +1,302 @@
+   We used a simple piece of Liquid markup to access our constants from the `context` object.
+
+```liquid
+{% raw %}{{context.constants.TWILIO_API_KEY_SID}}{% endraw %}
````
```suggestion
{% raw %}{{ context.constants.TWILIO_API_KEY_SID }}{% endraw %}
```

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+
+   **Assign Liquid timestamp variables**
+    
+   We needed to be able to tell Twilio the earliest possible time the token can be used and the time when the token should expire. We did this using [Liquid date filters](https://documentation.platformos.com/api-reference/liquid/filters#date): `date` for date format, `now` for time now and `add_to_time` to add the length of time before the token expires. 
```
```suggestion
   We needed to be able to tell Twilio the earliest possible time the token can be used and the time when the token should expire. We did this using [Liquid date filters](/api-reference/liquid/filters#date): `date` for date format, `now` for time now and `add_to_time` to add the length of time before the token expires. 
```

pavelloz commented on app/views/pages/use-cases/twilio-video-integration.liquid:
```diff
@@ -0,0 +1,302 @@
+
+### Step 4: Encode the JWT Token
+
+Now we had the variable and payload ready. We could encode payload to a **JWT** token using the [jwt_encode](https://documentation.platformos.com/api-reference/liquid/platformos-filters#jwt_encode) Liquid filter. To do this we needed to know the **Algorithm** to use for encryption (in this case `HS256`), the **Secret** constant we created earlier `context.constants.TWILIO_API_SECRET` and a **Cust ...
```
```suggestion
Now we had the variable and payload ready. We could encode payload to a **JWT** token using the [jwt_encode](/api-reference/liquid/platformos-filters#jwt_encode) Liquid filter. To do this we needed to know the **Algorithm** to use for encryption (in this case `HS256`), the **Secret** constant we created earlier `context.constants.TWILIO_API_SECRET` and a **Custom Header** of `{"cty": "twilio-fpa;v=1"}` where we again used `parse_json` to create a **Hash** from our JSON.
```

pavelloz reviewed (changes requested):
Good job.

Looks like a lot of research and hard work went into this.

Left couple suggestions/comments with typo fixes etc.

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Thanks for catching this stuff Pawel. Sorry for the delay on these it was a national holiday here. Went out walkabout with family.

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Sorry for this one. I should have checked. I changed my CLI to use `pos` when it was marketplace-kit so kept it.

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Update app/views/pages/use-cases/twilio-video-integration.liquid

Co-Authored-By: Paweł Kowalski <pavelloz@Gmail.com>

digitalfuel committed:
Removed charset

digitalfuel committed:
Add note to remove logs in production.

digitalfuel committed:
Added page path before code under 'Create a page'.

digitalfuel committed:
Removed second charset instance
